spec/dashboards/customer_dashboard_spec.rb

```diff
@@ -24,6 +24,45 @@
     end
   end
 
+  describe "#attribute_type_for" do
+    context "for an existing attribute" do
+      it "returns the attribute field" do
+        dashboard = CustomerDashboard.new
+        field = dashboard.attribute_type_for(:name)
+        expect(field).to eq Administrate::Field::String
+      end
+    end
+
+    context "for a non-existent attribute" do
+      it "raises an exception" do
+        dashboard = CustomerDashboard.new
+        expect { dashboard.attribute_type_for(:foo) }.
+          to raise_error /Attribute foo not in CustomerDashboard/
+      end
+    end
+  end
+
+  describe "#attribute_types_for" do
+    context "for existing attributes" do
+      it "returns the attribute fields" do
+        dashboard = CustomerDashboard.new
+        fields = dashboard.attribute_types_for(:name, :email)
+        expect(fields).to match(
+          name: Administrate::Field::String,
+          email: Administrate::Field::Email,
+        )
+      end
+    end
+
+    context "for one non-existent attribute" do
+      it "raises an exception" do
+        dashboard = CustomerDashboard.new
+        expect { dashboard.attribute_types_for(:name, :foo) }.
+          to raise_error /Attribute foo not in CustomerDashboard/
```

Comment:
How would you feel about extracting a method here?

```ruby
expect { dashboard.attribute_types_for(:name, :foo) }.
  to raise_error(message_for("foo", "CustomerDashboard"))
  
  
def message_for(attribute, dashboard_class)
  "Attribute #{attribute} could not be found in #{dashboard_class}::ATTRIBUTE_TYPES")
end
```

Note: I also slightly changed the wording of the error message in that snippet, I think it reads a bit better.